File: frontend/src/domain/component.ts
export type ComponentTemplate = {
  id: number | string
  component_key: string
  name: string
  name_en?: string
}

export type ComponentSchemaField = {
  key: string
  label: string
  type: 'text' | 'textarea' | 'number' | 'color' | 'switch' | 'select' | 'radio' | 'checkbox' | 'slider' | 'goods' | 'link' | 'list' | 'custom' | string
  default?: unknown
  required?: boolean
  tips?: string
  options?: Array<{ label: string; value: string | number | boolean }>
  min?: number
  max?: number
}

export type ComponentSchemaGroup = {
  key: string
  title: string
  fields: ComponentSchemaField[]
}

export type ComponentSchema = {
  groups?: ComponentSchemaGroup[]
}

export type ComponentMeta = {
  id: number | string
  component_key: string
  name: string
  category_id?: number | string
  icon?: string
  tpl_id?: number | string
  templates: ComponentTemplate[]
  schema_json?: ComponentSchema
  default_data?: Record<string, unknown>
  enabled?: 0 | 1
  sort?: number
}

export type DesignComponent = {
  id: string
  component_key: string
  component_title: string
  template_id: number | string
  remote_data: Record<string, unknown>
  tasks?: DesignComponent[]
  invalid?: boolean
  invalidReason?: string
}

File: backend/app/schemas/component.py
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ComponentTemplate(BaseModel):
    id: int | str
    component_key: str
    name: str
    name_en: str | None = None


class ComponentMeta(BaseModel):
    id: int | str
    component_key: str
    name: str
    category_id: int | str = 0
    icon: str | None = None
    tpl_id: int | str | None = None
    templates: list[ComponentTemplate] = Field(default_factory=list)
    schema_data: dict[str, Any] = Field(default_factory=dict, alias="schema_json")
    default_data: dict[str, Any] = Field(default_factory=dict)
    enabled: int = 1
    sort: int = 0

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class DesignComponent(BaseModel):
    id: str
    component_key: str
    component_title: str
    template_id: int | str
    remote_data: dict[str, Any] = Field(default_factory=dict)
    tasks: list["DesignComponent"] = Field(default_factory=list)
    invalid: bool | None = None
    invalidReason: str | None = None
